File: vane-proxy-core/src/main/java/org/oddlama/vane/proxycore/config/ManagedServer.java
package org.oddlama.vane.proxycore.config;

import com.electronwill.nightconfig.core.CommentedConfig;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.util.Base64;
import java.util.List;
import java.util.Random;
import javax.imageio.ImageIO;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ManagedServer {

    public String displayName;
    public ServerStart start;

    private final String id;
    private final StatefulConfiguration onlineConfig;
    private final StatefulConfiguration offlineConfig;

    public ManagedServer(
        String id,
        String displayName,
        CommentedConfig onlineConfigSection,
        CommentedConfig offlineConfigSection,
        CommentedConfig start
    ) throws IOException {
        this.id = id;
        this.displayName = displayName;

        this.onlineConfig = new StatefulConfiguration(id, displayName, onlineConfigSection);
        this.offlineConfig = new StatefulConfiguration(id, displayName, offlineConfigSection);
        this.start = new ServerStart(id, displayName, start);
    }

    public @NotNull String id() {
        return this.id;
    }

    public @Nullable String favicon(ConfigItemSource source) {
        switch (source) {
            case ONLINE -> {
                return onlineConfig.encodedFavicon;
            }
            case OFFLINE -> {
                return offlineConfig.encodedFavicon;
            }
            default -> {
                return null;
            }
        }
    }

    public String[] startCmd() {
        return start.cmd;
    }

    public String startKickMsg() {
        return start.kickMsg;
    }

    private String randomQuote(ConfigItemSource source) {
        final String[] quoteSet;
        switch (source) {
            case ONLINE -> quoteSet = onlineConfig.quotes;
            case OFFLINE -> quoteSet = offlineConfig.quotes;
            default -> {
                return "";
            }
        }

        if (quoteSet == null || quoteSet.length == 0) {
            return "";
        }
        return quoteSet[new Random().nextInt(quoteSet.length)];
    }

    public String motd(ConfigItemSource source) {
        final String sourcedMotd;
        final ConfigItemSource quoteSource;
        switch (source) {
            case ONLINE -> {
                sourcedMotd = onlineConfig.motd;
                quoteSource = ConfigItemSource.ONLINE;
            }
            case OFFLINE -> {
                sourcedMotd = offlineConfig.motd;
                quoteSource = ConfigItemSource.OFFLINE;
            }
            default -> {
                return "";
            }
        }

        if (sourcedMotd == null) {
            return "";
        }
        return sourcedMotd.replace("{QUOTE}", randomQuote(quoteSource));
    }

    public Integer commandTimeout() {
        return start.timeout;
    }

    public enum ConfigItemSource {
        ONLINE,
        OFFLINE,
    }

    private static class StatefulConfiguration {

        public String[] quotes = null;
        public String motd = null;
        private @Nullable String encodedFavicon;

        public StatefulConfiguration(String id, String displayName, CommentedConfig config) throws IOException {
            // [managedServers.my_server.state]
            if (config == null) {
                // The whole section is missing
                return;
            }

            // quotes = ["", ...]
            List<String> quotes = config.get("quotes");

            if (quotes != null) this.quotes = quotes
                .stream()
                .filter(s -> !s.isBlank())
                .map(s -> s.replace("{SERVER}", id).replace("{SERVER_DISPLAY_NAME}", displayName))
                .toArray(String[]::new);

            // motd = "..."
            var motd = config.get("motd");

            if (!(motd == null || motd instanceof String)) throw new IllegalArgumentException(
                "Managed server '" + id + "' has a non-string MOTD!"
            );

            if (motd != null && !((String) motd).isEmpty()) this.motd = ((String) motd).replace(
                    "{SERVER_DISPLAY_NAME}",
                    displayName
                );

            // favicon = "..."
            var faviconPath = config.get("favicon");

            if (!(faviconPath == null || faviconPath instanceof String)) throw new IllegalArgumentException(
                "Managed server '" + id + "' has a non-string favicon path!"
            );

            if (faviconPath != null && !((String) faviconPath).isEmpty()) this.encodedFavicon = encodeFavicon(
                id,
                (String) faviconPath
            );
        }

        private static String encodeFavicon(String id, String faviconPath) throws IOException {
            File faviconFile = new File(faviconPath.replace("{SERVER}", id));
            BufferedImage image;
            try {
                image = ImageIO.read(faviconFile);
            } catch (IOException e) {
                throw new IOException("Failed to read favicon file: " + e.getMessage());
            }

            if (image.getWidth() != 64 || image.getHeight() != 64) {
                throw new IllegalArgumentException("Favicon has invalid dimensions (must be 64x64)");
            }

            ByteArrayOutputStream stream = new ByteArrayOutputStream();
            ImageIO.write(image, "PNG", stream);
            byte[] faviconBytes = stream.toByteArray();

            String encodedFavicon = "data:image/png;base64," + Base64.getEncoder().encodeToString(faviconBytes);

            if (encodedFavicon.length() > Short.MAX_VALUE) {
                throw new IllegalArgumentException("Favicon file too large for server to process");
            }

            return encodedFavicon;
        }
    }

    public static class ServerStart {

        private static final int DEFAULT_TIMEOUT_SECONDS = 10;
        public String[] cmd;
        public Integer timeout;
        public String kickMsg;
        public boolean allowAnyone;

        public ServerStart(String id, String displayName, CommentedConfig config) {
            List<String> cmd = config.get("cmd");
            var timeout = config.get("timeout");
            var kickMsg = config.get("kickMsg");
            var allowAnyone = config.get("allowAnyone");

            if (cmd != null) this.cmd = cmd.stream().map(s -> s.replace("{SERVER}", id)).toArray(String[]::new);
            else this.cmd = null;

            if (!(kickMsg == null || kickMsg instanceof String)) throw new IllegalArgumentException(
                "Managed server '" + id + "' has an invalid kick message!"
            );

            if (kickMsg != null) this.kickMsg = ((String) kickMsg).replace("{SERVER}", id).replace(
                    "{SERVER_DISPLAY_NAME}",
                    displayName
                );
            else this.kickMsg = null;

            if (allowAnyone != null) this.allowAnyone = (boolean) allowAnyone;
            else this.allowAnyone = false;

            if (timeout == null) {
                this.timeout = DEFAULT_TIMEOUT_SECONDS;
                return;
            }

            if (!(timeout instanceof Integer)) throw new IllegalArgumentException(
                "Managed server '" + id + "' has an invalid command timeout!"
            );

            this.timeout = (Integer) timeout;
        }
    }
}

File: vane-proxy-core/src/main/kotlin/org/oddlama/vane/proxycore/config/ManagedServer.kt
package org.oddlama.vane.proxycore.config

import com.electronwill.nightconfig.core.CommentedConfig
import java.awt.image.BufferedImage
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.util.*
import javax.imageio.ImageIO

@Suppress("unused")
class ManagedServer(
    private val id: String,
    displayName: String,
    onlineConfigSection: CommentedConfig?,
    offlineConfigSection: CommentedConfig?,
    start: CommentedConfig
) {
    @JvmField
    var start: ServerStart = ServerStart(id, displayName, start)

    private val onlineConfig: StatefulConfiguration = StatefulConfiguration(id, displayName, onlineConfigSection)
    private val offlineConfig: StatefulConfiguration = StatefulConfiguration(id, displayName, offlineConfigSection)

    fun id(): String {
        return this.id
    }

    fun favicon(source: ConfigItemSource): String? {
        return when (source) {
            ConfigItemSource.ONLINE -> {
                onlineConfig.encodedFavicon
            }

            ConfigItemSource.OFFLINE -> {
                offlineConfig.encodedFavicon
            }

        }
    }

    fun startCmd(): Array<String?>? {
        return start.cmd
    }

    fun startKickMsg(): String? {
        return start.kickMsg
    }

    private fun randomQuote(source: ConfigItemSource): String? {
        val quoteSet = when (source) {
            ConfigItemSource.ONLINE -> onlineConfig.quotes
            ConfigItemSource.OFFLINE -> offlineConfig.quotes
        }

        if (quoteSet.isNullOrEmpty()) {
            return ""
        }
        return quoteSet[Random().nextInt(quoteSet.size)]
    }

    fun motd(source: ConfigItemSource): String {
        val sourcedMotd: String?
        val quoteSource: ConfigItemSource
        when (source) {
            ConfigItemSource.ONLINE -> {
                sourcedMotd = onlineConfig.motd
                quoteSource = ConfigItemSource.ONLINE
            }

            ConfigItemSource.OFFLINE -> {
                sourcedMotd = offlineConfig.motd
                quoteSource = ConfigItemSource.OFFLINE
            }

        }

        if (sourcedMotd == null) {
            return ""
        }
        return sourcedMotd.replace("{QUOTE}", randomQuote(quoteSource)!!)
    }

    fun commandTimeout(): Int? {
        return start.timeout
    }

    enum class ConfigItemSource {
        ONLINE,
        OFFLINE,
    }

    private class StatefulConfiguration(id: String, displayName: String, config: CommentedConfig?) {
        var quotes: Array<String?>? = null
        var motd: String? = null
        var encodedFavicon: String? = null

        init {
            // [managedServers.my_server.state]
            if (config != null) {
                // quotes = ["", ...]
                val quotesList = config.get<MutableList<String?>?>("quotes")

                if (quotesList != null) {
                    this.quotes = quotesList
                        .filter { s -> !s.isNullOrBlank() }
                        .map { s -> s!!.replace("{SERVER}", id).replace("{SERVER_DISPLAY_NAME}", displayName) }
                        .toTypedArray()
                }

                // motd = "..."
                val motdVal = config.get<Any?>("motd")

                require(motdVal == null || motdVal is String) { "Managed server '$id' has a non-string MOTD!" }

                if (motdVal is String && motdVal.isNotEmpty()) this.motd = motdVal.replace(
                    "{SERVER_DISPLAY_NAME}",
                    displayName
                )

                // favicon = "..."
                val faviconPath = config.get<Any?>("favicon")

                require(faviconPath == null || faviconPath is String) { "Managed server '$id' has a non-string favicon path!" }

                if (faviconPath is String && faviconPath.isNotEmpty()) this.encodedFavicon = encodeFavicon(
                    id,
                    faviconPath
                )
            }
        }

        companion object {
            @Throws(IOException::class)
            private fun encodeFavicon(id: String, faviconPath: String): String {
                val faviconFile = File(faviconPath.replace("{SERVER}", id))
                val image: BufferedImage
                try {
                    image = ImageIO.read(faviconFile)
                } catch (e: IOException) {
                    throw IOException("Failed to read favicon file: " + e.message)
                }

                require(!(image.width != 64 || image.height != 64)) { "Favicon has invalid dimensions (must be 64x64)" }

                val stream = ByteArrayOutputStream()
                ImageIO.write(image, "PNG", stream)
                val faviconBytes = stream.toByteArray()

                val encodedFavicon = "data:image/png;base64," + Base64.getEncoder().encodeToString(faviconBytes)

                require(encodedFavicon.length <= Short.MAX_VALUE) { "Favicon file too large for server to process" }

                return encodedFavicon
            }
        }
    }

    class ServerStart(id: String, displayName: String, config: CommentedConfig) {
        var cmd: Array<String?>?
        var timeout: Int?
        var kickMsg: String? = null
        @JvmField
        var allowAnyone: Boolean = false

        init {
            val cmdList = config.get<MutableList<String?>?>("cmd")
            val timeoutVal = config.get<Any?>("timeout")
            val kickMsgVal = config.get<Any?>("kickMsg")
            val allowAnyoneVal = config.get<Any?>("allowAnyone")

            if (cmdList != null) this.cmd = cmdList.map { s -> s!!.replace("{SERVER}", id) }.toTypedArray()
            else this.cmd = null

            require(kickMsgVal == null || kickMsgVal is String) { "Managed server '$id' has an invalid kick message!" }

            if (kickMsgVal is String) this.kickMsg = kickMsgVal.replace("{SERVER}", id).replace(
                "{SERVER_DISPLAY_NAME}",
                displayName
            )
            else this.kickMsg = null

            if (allowAnyoneVal is Boolean) this.allowAnyone = allowAnyoneVal
            else this.allowAnyone = false

            if (timeoutVal == null) {
                this.timeout = DEFAULT_TIMEOUT_SECONDS
            } else {
                require(timeoutVal is Int) { "Managed server '$id' has an invalid command timeout!" }

                this.timeout = timeoutVal
            }
        }

        companion object {
            private const val DEFAULT_TIMEOUT_SECONDS = 10
        }
    }
}
